core/src/main/java/org/jdbi/v3/core/statement/PreparedBatch.java

```diff
@@ -182,11 +181,15 @@ private ExecutedBatch internalBatchExecute() {
             beforeExecution(stmt);
 
             try {
+                getConfig(SqlStatements.class).getSqlLogger().logBeforeExecution(getContext());
+
                 final long start = System.nanoTime();
                 final int[] rs =  stmt.executeBatch();
                 final long elapsedTime = System.nanoTime() - start;
+
                 LOG.trace("Prepared batch of {} parts executed in {}ms", bindings.size(), elapsedTime / 1000000L, parsedSql);
                 getConfig(SqlStatements.class).getTimingCollector().collect(elapsedTime, getContext());
+                getConfig(SqlStatements.class).getSqlLogger().logAfterExecution(getContext(), elapsedTime);
```

Comment:
This pattern shows up a lot.  Would it make sense for SqlLogger to have a

```java
default void timedExecute(Runnable action, StatementContext ctx) throws SQLException {
    final SqlLogger logger = ctx.getConfig(SqlStatements.class).getSqlLogger();
    try {
        beforeExecution(ctx);
        long startNs = System.nanoTime();
        action.run();
        afterExecution(ctx, System.nanoTime() - startNs);
    } catch (SQLException | RuntimeException e) {
        logException(ctx, e);
        throw e;
    }
}
````

Reply:
I don't really understand...

Reply:
Steven is suggesting adding a function to `SqlLogger` (or just some static helper method somewhere), which encapsulates the logger boilerplate, instead of repeating it in multiple classes:

```java
interface  LogCallback<T, X extends Exception> {
  T invoke() throws X;
}

static <T, X extends Exception> T executeLogged(LogCallback<T, X> callback, StatementContext ctx) throws X {
  SqlLogger logger = ctx.getConfig(SqlStatements.class).getSqlLogger();
  try {
    ctx.setExecutionMoment(Instant.now());
    logger.logBeforeExecution(ctx);

    T result = callback.invoke();

    ctx.setCompletionMoment(Instant.now());
    logger.logAfterExecution(ctx);

    return result;
  } catch (RuntimeException | X e) {
    ctx.setExceptionMoment(Instant.now());
    logger.logException(ctx, e);
    throw e;
  }
}
```

Thus, in PreparedBatch:

```java

try {
    final int[] rs = SqlLoggers.executeLogged(stmt::executeBatch, getContext());

    afterExecution(stmt);

    getContext().setBinding(new Binding());

    return new ExecutedBatch(stmt, rs);
} catch (SQLException e) {
    throw new UnableToExecuteStatementException(Batch.mungeBatchException(e), getContext());
}
```

Reply:
Ah, now I get it. Hmm... I'll cook something up.

Reply:
Can't `catch X` though

Reply:
Catch `Exception` and unsafe-cast it to `X` to rethrow.

Reply:
See `LocalTransactionHandler.java` line 142 for an example.